### ISSUE TITLE
Wiring diagram?

### ISSUE THREAD

**Issue by multigamesystem** [DavidPagels/retro-pico-switch#36]
Could someone please provide a wiring diagram for wiring the GameCube controller connector to the Pico?

**Comment by CodeAsm** [comment]
based on:
> The pre-built .uf2 file assumes the data pin is on the Pico's [GP18 pin](https://datasheets.raspberrypi.com/pico/Pico-R3-A4-Pinout.pdf). A 1k pullup resistor should also be wired between 3.3v and your data pin to work properly.

I havent build this yet, but for gamecube, this might be how to. i made this based on the above hint from https://github.com/DavidPagels/retro-pico-switch?tab=readme-ov-file#setup using https://wokwi.com and a picture editor
[
![how](https://github.com/user-attachments/assets/239291fc-922c-43e6-b2a2-a51032bc0a41)
](url)

I hope you can figure out what pins of the connector go to where on the GC connector when you have one. if you snip it to wire directly, I have no idea what colors the cable have . https://simplecontrollers.com/blogs/resources/gamecube-protocol these numbers and layout I followed.
You could also move the resistor towards the pico connection wise (thus, solder from GP18 to 3.3v, and then attach the data wire of the GC)

**Comment by CodeAsm** [comment]
Turns out my N64 cable was dead broken, bypassed it and this wiring diagram is confirmed for atleast the N64 and its 3 wires. so whatever is power pin, ground pin and data, if you can figure those out, this schema works

**Comment by multigamesystem** [comment]
> Turns out my N64 cable was dead broken, bypassed it and this wiring diagram is confirmed for atleast the N64 and its 3 wires. so whatever is power pin, ground pin and data, if you can figure those out, this schema works

Thanks for your help I'm gonna attempt to wire up a test tomorrow.  I wonder if I'm able to get more than one controller input out of one Pico or do I have to use several to make a multi tap type input device?

**Comment by multigamesystem** [comment]
Eh it sorta works but Im getting all sorts of random inputs when not touching the gamepad.  Like some sort of noise on the data.  It seems counterintuitive to connect the data pin to 3.3V?

**Comment by CodeAsm** [comment]
> Eh it sorta works but Im getting all sorts of random inputs when not touching the gamepad. Like some sort of noise on the data. It seems counterintuitive to connect the data pin to 3.3V?

**TLDR: use the resistor between data and 3.3v, do not connect data directly to 3.3v. I dont think the pico can read more than 1 controller at a time.**

You must connect trough a resistor (dont connect to 3.3v directly, you may fry the pico). the resistor is acting as a "pull up" resistor. 
if the code would support mutliple data lines for multiple devices, each dataline should get its own resistor to 3.3v. this allows each dataline to pull down the line to ground. (its often easier for microcontrollers to pull to ground, then to pull to supplyvoltage)
https://en.wikipedia.org/wiki/Pull-up_resistor

some microcontrollers have build in resistors, or programmable options to do so. this might mean, similiar projects for different controllers may use the internal pull up or down resistors. 

I quickly looked with google images, and can not find multitap devices that seem simple, You probably need to alter the code to use another data line for another controller, issue 2 youll be having is now you also need to modify the usb code to present itself as a second usb-hid device (for the second controller). unless there is some usb device that supports the more buttons 2 controllers would pressent and the system you connect to support this bigger controller.

I assume you instead imagine connecting multiple N64 or GC controllers to for example a switch or PC as normal controllers, this project (sadly?) would suggest you get multiple pico and add the same hardware to convert each controller individually.
You may build these into a enclosure and use a usb hub, then you have 1 usb plug for your pc/switch (hope this works on the switch)

The Official controller  seems to use a IC that uses seperate data lines and probably exposes either multiple controllers or usb device with specifics that the console understands to interpret as multiple controllers. (I dont know how the official multitap usb device works or exposes the controllers on the usb side) 

EDIT: ![alt](https://bitbuilt.net/images/guide_hub/console_modding/gamecube/definitive_guide/omgwtf_dd.png "Title") from https://bitbuilt.net/forums/index.php?threads/the-definitive-gc-motherboard-trimming-guide.99/#post-403/
here you can see on the original GC, the 4 player data go trough a transisitor into the FLIPPER gpu ic (thus, the gpu handles the input seperatly, no combination logic to be found.

**Comment by multigamesystem** [comment]
> > Eh it sorta works but Im getting all sorts of random inputs when not touching the gamepad. Like some sort of noise on the data. It seems counterintuitive to connect the data pin to 3.3V?
> 
> **TLDR: use the resistor between data and 3.3v, do not connect data directly to 3.3v. I dont think the pico can read more than 1 controller at a time.**
> 
> You must connect trough a resistor (dont connect to 3.3v directly, you may fry the pico). the resistor is acting as a "pull up" resistor. if the code would support mutliple data lines for multiple devices, each dataline should get its own resistor to 3.3v. this allows each dataline to pull down the line to ground. (its often easier for microcontrollers to pull to ground, then to pull to supplyvoltage) https://en.wikipedia.org/wiki/Pull-up_resistor

Have you actually wired up a single Official GameCube controller using this PICO program and wiring diagram?  Could you test?  I cant get it to work.

**Comment by multigamesystem** [comment]
Ive designed a casing for the USB-C version of the Pico.  I already did a test print with my resin 3D printer.  It came out good.  I also just designed a little PCB that solders to the end of the GameCube port jack to make soldering wires and the resistor easier.  I want to make sure the wiring all makes sense before I commit to having these little PCBs made.

![Image](https://github.com/user-attachments/assets/67419493-b3b2-40bd-beec-e3552af56b70)
![Image](https://github.com/user-attachments/assets/c02ce9ef-813f-4872-a603-b9b05021349c)

**Comment by multigamesystem** [comment]
Also designed a 4 port version with USB hub inside to handle the 4 inputs.  While this seems wasteful and maybe expensive to achieve something that's already being done with a single soc and pcb.  I could make my own RP2040 board but thats maybe beyond my PCB skillset.  

![Image](https://github.com/user-attachments/assets/d66b459a-63b0-4f50-9e33-fab6b0aa3bca)
![Image](https://github.com/user-attachments/assets/1ddc55e8-fc70-4026-9803-a29b6dd35fd5)
![Image](https://github.com/user-attachments/assets/5b0f813a-7bb5-43c5-9279-395ccf05e879)

**Comment by multigamesystem** [comment]
Details of my GC Port add on pcb.  I hope it makes sense how the wiring is setup.  I used your guide as reference.  

![Image](https://github.com/user-attachments/assets/c07edf88-23c8-4056-a947-d3ce8edfa010)
![Image](https://github.com/user-attachments/assets/04f325c7-171e-4f59-849f-c01505c84a31)

![Image](https://github.com/user-attachments/assets/095694ec-f5a6-4dc7-bdfc-9eb94613fc38)

**Comment by multigamesystem** [comment]
Finally got it working I plan on making an updated wiring diagram.  Hopefully my PCBs work as intended when they arrive. 

![Image](https://github.com/user-attachments/assets/2701c624-9261-448a-9f5c-3b8f84afcc62)

**Comment by CodeAsm** [comment]
looks cool :D ive only used a N64 controller (I bought "broken, no refund" from a thrift store). for your PCB skills, look awesome :D
If you go for double sided (current design could be done at home on a single sided pcb), ive seen often they pick 1 side as a whole ground and other as main supply voltage. but whatever works and with this simple circuit, you cant go wrong I think.

Wish I had a 3d printer (but instead, we went on a vacation, just returned, hence the delay in reply). If you use a differnt internals with the GC shell, you could route the usb trough one of the GC SIO(expansion?) ports maybe? you may even be able to tap into one of those USB-B plugs and use another usb hub to keep that one working. 
(chaining is possible, but at a certain point, you do need to provide "external" power). Just ideas, awesome how your building 😄

**Comment by harrycb7** [comment]
Hey guys. Have either of you worked out where to wire the pins 1 and 4 to the Pico for rumble support?
